### ISSUE TITLE
Question on maskcat mutate

### ISSUE THREAD

**Issue by kart61** [JakeWnuk/maskcat#2]
Hello.
I am mastering your tools. Very useful tools!
I found a strange (possibly incorrect) work when using "maskcat mutate".

Even when executing your example:
$ cat test.txt
this
is a
Test!
love
Testing
Work$

$ cat test.txt | shuf | maskcat mutate 4
Work$
Test$
Work!
Test!
Tthisng
Working
Testing
Tloveng

In your example, the end result of executing "maskcat mutate 4" contains 8 lines. 
I had to execute "maskcat mutate 4" several times to get this result. 
Very often the result contains only 5, 6 or 7 lines out of 8 possible lines.

The same problem is observed when I used my own dataset as well.
I could not find any regularity as to why this happens.

Perhaps there are some nuances when running this command?

**Comment by JakeWnuk** [comment]
Hello! Thanks for opening an issue, this is probably a common question for users so for some additional details:

- The way `mutate` currently works is the additional parameter for `maskcat mutate <INT>` would change the length of the tokens kept by the parser and not the final output length. 
- This can be found in `internal/cli.go:205` ([link](https://github.com/JakeWnuk/maskcat/blob/034adbee126d59b6144a1ca70839c50886214452/internal/cli/cli.go#L205)), and can be changed by altering the logic of `if len(token) >= chunksInt { ...`.
- The reason for this is mostly personal preference, as I was writing the tool sometimes the chunk size of the token would need to be specific in order to extract certain patterns. Tokens of length 1-2 also could "wipe out" the string causing an entire string to be replaced by a single character or two. This issue was fixed when the `-n` parameter was introduced to control the maximum number of replacements per item to prevent too many replacements. 

It is a bit of a nuance of the `mutate` mode and is referenced in `docs/SWAP_AND_MUTATE.md` ([link](https://github.com/JakeWnuk/maskcat/blob/main/docs/SWAP_AND_MUTATE.md)) with little detail on what it means. 
- "If the token length is greater than or equal to the input length then it is kept otherwise discarded and not used for swapping."

So for the second example with `cat test.txt | shuf | maskcat mutate 4`, the text is parsed for tokens and anything with a length of four is kept. This would likely be `love`, `this`, `Test`, and `Work` which is why they primarily show up in the results.

The goal of the tool is to provide as many unique ways of isolating and transforming patterns as possible so I would be open to any community suggestions on what the default behavior should be as well. Hopefully this provided some clarity on what the tool is doing!

**Comment by kart61** [comment]
Thanks for the quick response and clarification of the nuances!

There are no comments on the logic of the program (the documentation also looked carefully), it works correctly.
Perhaps I did not describe the problem correctly, it is associated with the output of the results.

We have an input file:
$ cat test.txt
This
IS a
Test!
Love
Testing
Work $

Example file after execution:
$ cat test.txt | Shuf | Maskcat Mutate 4
Work $
Test $
Work!
Test!
Tthisng
Working
Testing
Tloveng
Everything is completely correct, predictably issues 8 lines of the result that are possible.

But if you execute this command "$ cat test.txt | shuf | maskcat mutate 4", then the results are:
Example 1:
Test!
Work!
Work $
Testing
Working
Tthisng
Tloveng
One result was lost ...

Example 2:
Tthisng
Test $
Work $
Test!
Work!
Three results were lost ...
And so on...

Since the input sequence does not change, the output result should be the same - but it turns out that the result is often from 3 to 7 lines of the output sequence, and it should always be 8.
Can there be any mistake here?

**Comment by JakeWnuk** [comment]
Ahhh yes I think I understand the issue better. 

So I looked over the documentation and it seems I removed a key point in the `mutate` mode that used to be in the documentation. So the `mutate` mode is super interesting because it is non-deterministic by design. 

This means that it will not generate identical output for the same input. The two reasons this happens is:
- Traditionally by the `shuf` command in example
- Newly by the concurrency provided by the goroutines.

What is happening is that the items are being read one-by-one from standard input and goes through the following steps at a high level:
- Parse text for ngrams/tokens and add them to the map
- Iterate over the map of tokens and perform swaps on possible tokens

The input is not changing but what is changing is the order they are processed in. For example, if "Test123" goes in and the ngram "Test" is pulled out and added to the map. Next, "Work345" is entered and the token "Work" is pulled out and added to the map. When it comes time for the candidates to be made for "Work345" the map contains "Test" and "Work" making the possible options "Test345" and "Work345". However, if the order was reversed and "Work345" was processed first and "Test123" was processed after then "Work123" and "Test123" would be the output.

Once the program is started, the map begins to fill with different items and depending on the order in which they are processed the output could be different. This is also multiplied by using the `shuf` command to mix up in the input and goroutines will also process items in a different order due to the multiple "threads" being used.

Let me know if this describes the behavior you are seeing and if there is anything else that is not quite clear from the documentation. These questions are really helpful to improve the documentation and help others who may be experiencing the same issues.
